FWCore/Framework/src/StreamSchedule.cc

```diff
@@ -425,23 +427,85 @@ namespace edm {
     for (auto w : allWorkers()) {
+      if (modulesToExclude.end() != modulesToExclude.find(w->description()->moduleLabel())) {
+        continue;
+      }
       //determine if this module could read a branch we want to delete early
-      auto pset = pset::Registry::instance()->getMapped(w->description()->parameterSetID());
-      if (nullptr != pset) {
-        auto branches = pset->getUntrackedParameter<std::vector<std::string>>("mightGet", kEmpty);
-        if (not branches.empty()) {
-          ++upperLimitOnReadingWorker;
-        }
-        for (auto const& branch : branches) {
-          auto found = branchToReadingWorker.equal_range(branch);
-          if (found.first != found.second) {
-            ++upperLimitOnIndicies;
-            ++reserveSizeForWorker[w];
-            if (nullptr == found.first->second) {
-              found.first->second = w;
+      auto consumes = w->consumesInfo();
+      if (not consumes.empty()) {
+        bool foundAtLeastOneMatchingBranch = false;
+        for (auto const& product : consumes) {
+          std::string branch = product.type().friendlyClassName() + "_" + product.label().data() + "_" +
+                               product.instance().data() + "_" + product.process().data();
+          {
+            //Handle case where worker directly consumes product
+            auto found = branchToReadingWorker.end();
+            if (product.process().empty()) {
+              auto startFound = branchToReadingWorker.lower_bound(branch);
+              if (startFound != branchToReadingWorker.end()) {
+                if (startFound->first.substr(0, branch.size()) == branch) {
+                  //match all processNames here, even if it means multiple matches will happen
+                  found = startFound;
+                }
+              }
+            } else {
+              auto exactFound = branchToReadingWorker.equal_range(branch);
+              if (exactFound.first != exactFound.second) {
+                found = exactFound.first;
+              }
+            }
+            if (found != branchToReadingWorker.end()) {
+              if (not foundAtLeastOneMatchingBranch) {
+                ++upperLimitOnReadingWorker;
+                foundAtLeastOneMatchingBranch = true;
+              }
+              ++upperLimitOnIndicies;
+              ++reserveSizeForWorker[w];
+              if (nullptr == found->second) {
+                found->second = w;
+              } else {
+                branchToReadingWorker.insert(make_pair(found->first, w));
+              }
+            }
+          }
+          {
+            //Handle case where indirectly consumes product
+            auto found = referencesToBranches.end();
+            if (product.process().empty()) {
+              auto startFound = referencesToBranches.lower_bound(branch);
+              if (startFound != referencesToBranches.end()) {
+                if (startFound->first.substr(0, branch.size()) == branch) {
```

Comment:
I don't think it is that simple. `std::string_view` can not be directly compared with a `std::string` so both sides would have to be converted to `std::string_view`. Doing a call like
```
std::string_view(startFound->first, branch.size())` isn't safe since `first` could be shorter than `branch.size()` and doing such is explicitly undefined behavior.